gnplt comment handling clean-up

wehimwich opened the pull request:
This PR started out as an answer to issue #111 (which has more discussion).  It covers three _gnplt_ comment related changes:

- Not double copying trailing comments in the .rxg files when updating.
- Not adding a _leading_ `RXG file updated by GnPlt2` date comment. It is no longer needed now that the date line is being updated.
- A script, _rxgfix3_, to help clean-up the _leading_ `GnPlt2` date comments that were added by _gnplt_.

jfhquick reviewed (commented):
Technically the removal of the leading comments and the corresponding script belongs with the earlier change that updated the date line of the rxg.  The fix for double copying of trailing comments should be implemented on its own.
I don't know enough perl to properly review the script but it looks sane and seems to work in the limited testing I have done.

wehimwich commented:
Thanks for the feedback and testing!

Sorry, I really find _perl_ expressive for small text processing jobs this. Plus, there is a collection of scripts like this with the infrastructure already built-up.

The changes in _gnplt_ are disjoint now as you say. The fix to the date line was a long time ago.  Fortunately (perhaps), no longer adding the date comment line was not implemented at the same time. That allows the possibility for a script to fix the trailing comments for _.rxg_ files that have only been updated by _gnplt_, i.e., not hand edited (see #111). Unfortunately, installing these changes in _gnplt_ will remove that possibility going forward.  So the fix to the trailing comments should really be made when the change is installed. It could be a lot of work to do that. That is probably not appropriate for a patch release, which would delay the changes for those who have worked around the bug by editing their trailing comments (and would like to stop having to do that!).

Perhaps a compromise is to add an environment variable to disable these changes in _gnplt_.  The patch release could recommend setting the environment variable until the user understands the implication of enabling the changes. That is sort of backwards since it would be better to be backward compatible with no action. However, it would also be better to not require setting an environment variable for the preferred long term behavior. Other opinions are welcome.

The patch release could ask the users to report if they are interested in using the proposed script to fix-up the trailing comments. If enough interest is expressed, the script could be added to a future release. That way the work of implementing the script is deferred until it is clear that it is worthwhile.

BTW, I squashed the three _rxgfix3_ commits down to one. This will impact you if you pull this branch again.

jfhquick commented:
Since rxgfix3 only strips the comments from the top of the file. all of the older copies and related comments are still intact lower down so reconstruction is still possible.  In the end it's only a matter of collecting all old copies, doing a unique sort into the correct reverse date order and overwriting all the original trailing comments with a clean copy. Just saying...

wehimwich commented:
I am concerned when you are just saying.

This issue has had a bit of an odyssey, from thinking there is an easy fix up to understanding that the fix-up is complicated and will only work for certain cases. A simpler proposal as you suggest would be a good reset.

It is marginally harder to fix the trailing comments if _rxgfix3_ has been run. It is also marginally harder to fix them if there have been updates with the improved _gnplt_. Neither case is impossible. The easiest case for writing a script to fix it is if the files only have the previous (bad) comment structure. However, we don't know how much interest there would be in that script.

So a simple(r) alternative is:

1. No environment variable (thanks).
2. Fixing the comments is entirely optional.
3. Consider running _rxgfix3_
4. If you want/need to fix your trailing comments:
- consider making a back-up of the current _.rxg_ files in case you want to fix them later. 
- consider deleting your current trailing comments after you have backed them up.
- description of how to fix comments for the simplest Istill not very simple) case, mostly from #111.

jfhquick commented:
I think your simple solution would be the way to go, but with one improvement.  Make _rxgfix3_ also remove all the trailing comments from the files it is asked to fix.  It makes the necessary backups and the user can reconstruct a clean history from them if they so desire.

wehimwich commented:
Updated _rxgfix3_:

- I added an option, `-c`, to cut trailing comments (kind of unperlish, but good enough).
- I added a prepass to make sure the files were available and there were no already existing _.bak_ files. There is also now an option, `-d`, to delete the latter. I thought this might be helpful since It can be a pain to recover if  you are processing a bunch of files and it gives up part way through the processing because of existing _.bak_ files. That could be useful in future utilities like this too.

There could be an option to make a back-up of the existing _.rxg_ files, but maybe that is best handled manually. It might be nice to have more verbose output, so the user can see if the changes make sense, but this is just a one time use utility and the `-t `option can be used to preview the changes.

jfhquick commented:
The new _rxgfix3_ functionality seems to work in my simple testing.  My only remaining comment is that the help text could do with some mention of why you might want to cut all trailing comments.  Otherwise it all looks good.

wehimwich commented:
Finally got around to this. I add the `-c` explanation. It is covered in the update notes as well. I have a rough draft in the notes of how to clean-up the trailing comments. It is a pretty laborious. I point out that it is not necessary and can be deferred. In some sense you will be just as well off (less pain, so maybe better), if you don't do anything.

jfhquick commented on misc/rxgfix3:
```diff
@@ -89,19 +89,22 @@ FILE: foreach $name (@ARGV) {
-    print "  You can delete them as part of the processing by adding the -d option.\n";
+    print
+"  Some back-up file(s), see messages above, already exist or couldn't
+  be deleted. Unless there an error deleting them, you can delete them
```
"unless there ?is?was? an error" perhaps

wehimwich replied:
:) "is" I guess. This is the most complete one-time fix-up utility we have ever had. Hopefully, it will be useful.
